### PR TITLE
Added source code part of ThankQ integration.

### PR DESCRIPTION
**Pull request by spleshka**

### CODE REVIEW

File: modules/features/falcon_feature_user/config/install/user.role.appeals_manager.yml [hunk 1]
```diff
@@ -12,15 +12,19 @@ permissions:
   - 'access toolbar'
   - 'create appeal content'
   - 'create media'
+  - 'create source_code entities'
```
Comment (kalabro) [inline modules/features/falcon_feature_user/config/install/user.role.appeals_manager.yml]: Why would source code permissions go into falcon_feature_user? Correct me if I'm wrong but I thought it should go to falcon_feature_thankq.

File: modules/features/falcon_feature_user/falcon_feature_user.info.yml [hunk 2]
```diff
@@ -5,6 +5,7 @@ core: 8.x
 dependencies:
   - field
   - menu_link_config
+  - path
   - user
```
Comment (kalabro) [inline modules/features/falcon_feature_user/falcon_feature_user.info.yml]: Why do we need path module here?

File: modules/features/falcon_feature_thankq/config/install/core.entity_form_display.source_code.source_code.default.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      size: 30
+      placeholder: 'I saw an ad on TV'
+    third_party_settings: {  }
+    type: string_textfield
+    region: content
+  field_source_code:
+    weight: 3
+    settings:
+      size: 10
+      placeholder: ASDR18-10
```
Comment (kalabro) [inline modules/features/falcon_feature_thankq/config/install/core.entity_form_display.source_code.source_code.default.yml]: I'd avoid real values in placeholders. For me **personally**, it's a bit confusing. I'd either remove the placeholder or make it something like "XXXX00-00"

**Review by kalabro — CHANGES_REQUESTED** [review]

File: modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+
+    // Mainly use parent code to render the form.
+    $element = parent::formElement($items, $delta, $element, $form, $form_state);
+
+    // Change field label to the one related to source codes.
+    $element['entities']['#table_fields']['label']['label'] = t('Source code');
```
Comment (duozersk) [inline modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php]: We have $this->t() available here.

File: modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+      // Grab source code & human readable label.
+      $source_code = $entity->get('field_source_code')->getString();
+      $label = $entity->get('field_label')->getString();
+      $date = '';
+
+      // Render availability date in the human readable format.
+      $availability = $entity->get('field_availability')->getValue();
+      if (!empty($availability[0]['value'])) {
+        $date .= format_string('from @from ', ['@from' => $availability[0]['value']]);
```
Comment (duozersk) [inline modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php]: format_string() is deprecated - https://api.drupal.org/api/drupal/core%21includes%21bootstrap.inc/function/format_string/8.5.x - looks like we need to use https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Render%21FormattableMarkup.php/function/FormattableMarkup%3A%3AplaceholderFormat/8.5.x

File: modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    foreach (Element::children($element['entities']) as $key) {
+
+      // Get Source Code entity.
+      /* @var $entity \Drupal\eck\EckEntityInterface */
+      $entity = $element['entities'][$key]['#entity'];
+
+      // Grab source code & human readable label.
+      $source_code = $entity->get('field_source_code')->getString();
+      $label = $entity->get('field_label')->getString();
```
Comment (duozersk) [inline modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php]: Looks unusual to me... Usually it is enough to write smth like `$entity->field_label->value`.

File: modules/custom/falcon_thankq/falcon_thankq.info.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+name: Falcon ThankQ
+type: module
+description: ThankQ integration and modifications for Falcon.
+core: 8.x
+package: Falcon
```
Comment (duozersk) [inline modules/custom/falcon_thankq/falcon_thankq.info.yml:5]: Shouldn't we have dependency on `inline_entity_form` here?

File: modules/features/falcon_feature_user/config/install/user.role.appeals_manager.yml (hunk 1 above)
Comment (duozersk) [reply modules/features/falcon_feature_user/config/install/user.role.appeals_manager.yml]: It does trigger me too... only if there is some issue with having permissions that are not defined (due to some modules not installed as chosen by client project) inside the falcon_feature_user - only in this case I would go with the roles permissions overriding in this feature... and still it means we would need to update this config here every time we change permissions (along with falcon_feature_user configs) - doesn't look maintainable at all...

**Review by duozersk — CHANGES_REQUESTED** [review]

File: modules/features/falcon_feature_user/config/install/user.role.appeals_manager.yml (hunk 1 above)
Comment (spleshka) [reply modules/features/falcon_feature_user/config/install/user.role.appeals_manager.yml]: @kalabro it's a bug, thanks, fixed.

File: modules/features/falcon_feature_user/falcon_feature_user.info.yml (hunk 2 above)
Comment (spleshka) [reply modules/features/falcon_feature_user/falcon_feature_user.info.yml]: @kalabro auto-dependency. But I agree there is no need, fixed.

File: modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php (hunk 4 above)
Comment (spleshka) [reply modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php]: @duozersk thanks, fixed.

File: modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php (hunk 5 above)
Comment (spleshka) [reply modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php]: @duozersk agree, fixed.

File: modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php (hunk 6 above)
Comment (spleshka) [reply modules/custom/falcon_thankq/src/Plugin/Field/FieldWidget/InlineSourceCodeForm.php]: @duozersk I don't have a strong preference, happy with your version cause it's more minimalistic.

File: modules/custom/falcon_thankq/falcon_thankq.info.yml (hunk 7 above)
Comment (spleshka) [reply modules/custom/falcon_thankq/falcon_thankq.info.yml:5]: @duozersk great catch, indeed.

**Review by kalabro — APPROVED** [review]

**Review by duozersk — APPROVED** [review]

**Pull request merged**